### PR TITLE
[MCPE] v1.8-1.10

### PR DESCRIPTION
**Pull request by colinrgodsey**
* UDP query listener on same stream as PE
* bungee-driven dimension switching (in a way that doesnt crash or confuse the PE client)
* custom event packet support

### CODE REVIEW

File: src/protocolsupport/protocol/packet/handler/PSInitialHandler.java [hunk 1]
```diff
@@ -218,7 +218,7 @@ protected void processLoginStart() {
 				return;
 			}
 			case PE: {
-				if (isOnlineMode()) {
+				if (true || isOnlineMode()) {
```
Comment (Shevchik) [inline src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: ?

Comment (MrPowerGamerBR) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: I think I made a fucky wucky on my branch, I added this because for some reason authentication wasn't working, let me disable that.

Comment (MrPowerGamerBR) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: Actually, it seems @colinrgodsey already removed my mistake

Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: ohhh that was you lol. i thought that was me

Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: @MrPowerGamerBR i dont even think i looked what i pulled over from that branch.. just seemed like the most up-to-date code 🤷‍♂️

File: src/protocolsupport/injector/pe/PEProxyServerInfoHandler.java [hunk 2]
```diff
@@ -53,7 +53,7 @@ public String getServerInfo(Channel channel) {
 			return String.join(";",
 				"MCPE",
 				ping.getMotd().toLegacyText().replace(";", ":"),
-				String.valueOf(ping.getVersion().getName()), "1.12.13",
+				String.valueOf(ping.getVersion().getName()), "1.7.0",
```
Comment (7kasper) [inline src/protocolsupport/injector/pe/PEProxyServerInfoHandler.java]: Can you do this like in PS main? https://github.com/ProtocolSupport/ProtocolSupport/blob/90c738615ddb9c898735f3ae9360baad031c9e4b/src/protocolsupport/zplatform/impl/pe/PEProxyServerInfoHandler.java#L64

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java [hunk 3]
```diff
@@ -53,14 +56,8 @@ protected void read0(ByteBuf from) {
 		from.readBoolean(); //broadcast to xbl
 		from.readBoolean(); //commands enabled
 		from.readBoolean(); //needs texture pack
-		int gameRuleCount = VarNumberSerializer.readVarInt(from); //game rules
```
Comment (7kasper) [inline src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: Why was this removed?

Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: was removed from the PSPE support and always reports 0 now:
https://github.com/ProtocolSupport/ProtocolSupport/commit/cd7cbb61acd1262a7085035b120ad8908996e07d#diff-226b405d338e937694734097e3342101R74

File: src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java [hunk 4]
```diff
@@ -27,7 +28,7 @@
 		registry.register(Protocol.GAME, PEPacketId.Clientbound.PLAY_KICK, KickPacket.class);
 		registry.register(Protocol.GAME, PEPacketId.Clientbound.PLAY_START_GAME, LoginPacket.class);
 		registry.register(Protocol.GAME, PEPacketId.Dualbound.PLAY_CHAT, FromServerChatPacket.class);
-		registry.register(Protocol.GAME, PEPacketId.Clientbound.PLAY_RESPAWN, RespawnPacket.class);
```
Comment (colinrgodsey) [inline src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java]: although im not sure why this was removed

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java (hunk 3 above)
Comment (Shevchik) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: While it does report 0, that zero still needs to be read from data.

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java [hunk 5]
```diff
@@ -53,14 +56,8 @@ protected void read0(ByteBuf from) {
...
-		for (int i = 0; i < gameRuleCount; i++) {
-			StringSerializer.readVarIntUTF8String(from); //game rule name
-		}
-		int gameRuleBoolCount = VarNumberSerializer.readVarInt(from); //game rules bool
-		for (int i = 0; i < gameRuleBoolCount; i++) {
-			from.readBoolean();
-		}
+		VarNumberSerializer.readVarInt(from); //game rule count
```
Comment (colinrgodsey) [inline src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: i still read the count here, just gets discarded

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java (hunk 3 above)
Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: still read it right down here:
https://github.com/ProtocolSupport/ProtocolSupportBungee/pull/69/files/1e7d2237011cf09863d12a04d3dce5ad98e274df#diff-10a180b559675d36c25de12bd641660dR59

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/RespawnPacket.java [hunk 6]
```diff
@@ -20,6 +20,18 @@ public RespawnPacket() {
 
 	protected int dimensionId;
 
+	/*
```
Comment (colinrgodsey) [inline src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/RespawnPacket.java]: removing this too

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java [hunk 7]
```diff
@@ -72,12 +69,14 @@ protected void read0(ByteBuf from) {
...
 		from.readBoolean(); // hasLockedBeh pack
 		from.readBoolean(); // hasLocked world template.
+		from.readBoolean(); // Microsoft GamerTags only. Hell no!
 		StringSerializer.readVarIntUTF8String(from); //level id (pe one)
 		StringSerializer.readVarIntUTF8String(from); //level name (will packet.getLevelType() work?)
 		StringSerializer.readVarIntUTF8String(from); //template pack id
 		from.readBoolean(); //is trial
 		from.readLongLE(); //level time
 		VarNumberSerializer.readSVarInt(from); //enchantment seed
+		PE_RUNTIME_IDS = MiscSerializer.readAllBytes(from); //also consumes version name
```
[the remark this answers is not in the document]
Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: ill have to dig around bungee a bit more, probably some way to keep a cache somewhere

[the remark this answers is not in the document]
Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: awesome, theres already cache object here. just used that

File: src/protocolsupport/protocol/packet/handler/PSInitialHandler.java [hunk 8]
```diff
@@ -218,7 +218,7 @@ protected void processLoginStart() {
 				return;
 			}
 			case PE: {
-				if (isOnlineMode()) {
+				if (BungeeCord.getInstance().config.isOnlineMode() || isOnlineMode()) {
```
Comment (Shevchik) [inline src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: Not again

**Comment by MrPowerGamerBR** [comment]
About the offline thingy, I *think* I added the "always true" value because, if I recall correctly, it was having some issues when decoding the chain.

However it was so long ago that I don't recall what was the exact error.

File: src/protocolsupport/protocol/packet/handler/PSInitialHandler.java (hunk 8 above)
Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: hmm dont we need this if bungee is in online mode though? isOnlineMode ends up getting set to false because the server connection isnt in offline mode =\

Comment (Shevchik) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: It was getting false because of a bug.

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java (hunk 3 above)
Comment (7kasper) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: Yeah I removed it from PSPE because I couldn't get it to work yet. But if we reimplement it in PE we need to do it here too. I think the encoding was right, I possibly had wrong names or so. Anyway if it works we might as well keep it the way it was. Otherwise this is fine for now.

File: src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java (hunk 4 above)
Comment (7kasper) [reply src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java]: neither do I... @MrPowerGamerBR ?

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (7kasper) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: 👍

File: src/protocolsupport/protocol/packet/handler/PSInitialHandler.java (hunk 8 above)
Comment (7kasper) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: Shev fixed this now

File: src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java (hunk 4 above)
Comment (MrPowerGamerBR) [reply src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java]: 🤷‍♀️ 

I think I disabled it while testing and forgot to reenable.

Comment (colinrgodsey) [reply src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java]: i put it back and test, seems fine?

File: src/protocolsupport/protocol/packet/handler/PSInitialHandler.java (hunk 8 above)
Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/handler/PSInitialHandler.java]: awesome, but it back to just isOnlineMode, works great now!

File: src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java (hunk 3 above)
Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/middleimpl/readable/play/v_pe/LoginPacket.java]: looks the game rules are statically typed now?
https://github.com/NukkitX/Nukkit/blob/master/src/main/java/cn/nukkit/level/GameRules.java

File: src/protocolsupport/protocol/pipeline/version/v_pe/FromClientPacketDecoder.java [hunk 9]
```diff
@@ -58,10 +58,13 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buf, List<Object> packe
 			return;
 		}
 		buf.markReaderIndex();
-		ReadableMiddlePacket transformer = registry.getTransformer(protocol, PEPacketIdSerializer.readPacketId(buf), false);
+		int packetId = PEPacketIdSerializer.readPacketId(buf);
+		//System.out.println("FROM CLIENT: " + packetId);
```
Comment (7kasper) [inline src/protocolsupport/protocol/pipeline/version/v_pe/FromClientPacketDecoder.java]: Remove this

File: src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java [hunk 10]
```diff
@@ -50,10 +51,13 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buf, List<Object> packe
 			return;
 		}
 		buf.markReaderIndex();
-		ReadableMiddlePacket transformer = registry.getTransformer(Protocol.GAME, PEPacketIdSerializer.readPacketId(buf), false);
+		int packetId = PEPacketIdSerializer.readPacketId(buf);
+		//System.out.println("FROM SERVER: " + packetId);
```
Comment (7kasper) [inline src/protocolsupport/protocol/pipeline/version/v_pe/FromServerPacketDecoder.java]: Remove this..

**Comment by 7kasper** [comment]
Ok just the debug and I think its good! 👍

File: src/protocolsupport/protocol/packet/handler/EntityRewriteDownstreamBridge.java [hunk 11]
```diff
@@ -28,4 +29,12 @@ public void handle(PacketWrapper packet) throws Exception {
 		con.sendPacket(packet);
 	}
 
+	@Override
```
Comment (colinrgodsey) [inline src/protocolsupport/protocol/packet/handler/EntityRewriteDownstreamBridge.java]: hmm ill move this

Comment (7kasper) [reply src/protocolsupport/protocol/packet/handler/EntityRewriteDownstreamBridge.java]: 😕

Comment (colinrgodsey) [reply src/protocolsupport/protocol/packet/handler/EntityRewriteDownstreamBridge.java]: i still cant find how/where to move this

Comment (7kasper) [reply src/protocolsupport/protocol/packet/handler/EntityRewriteDownstreamBridge.java]: Can you check if this is still needed after we merge master into PSPE?

File: src/protocolsupport/protocol/packet/handler/EntityRewriteDownstreamBridge.java [hunk 12]
```diff
@@ -28,4 +29,12 @@ public void handle(PacketWrapper packet) throws Exception {
...
+	public void handle(PluginMessage pluginMessage) throws Exception {
+		// bungee somehow doesnt update these, even though our protocol technically supports the new format
+		if(pluginMessage.getTag().equals("bungeecord:main")) {
```
Comment (7kasper) [inline src/protocolsupport/protocol/packet/handler/EntityRewriteDownstreamBridge.java]: Space

File: src/protocolsupport/protocol/pipeline/version/v_pe/ToClientPacketEncoder.java [hunk 13]
```diff
@@ -48,10 +55,40 @@
...
 		registry.register(BossBar.class, NoopWriteablePacket.class);
 		registry.register(Title.class, NoopWriteablePacket.class);
+		registry.register(PluginMessage.class, CustomEventPacket.class);
 	}
 
 	public ToClientPacketEncoder(Connection connection, NetworkDataCache cache) {
 		super(connection, cache);
 	}
 
+	// wish java had some kind of real 'tuple' class
```
Comment (7kasper) [inline src/protocolsupport/protocol/pipeline/version/v_pe/ToClientPacketEncoder.java]: ProtocolSupport has a touble class wich you can copy over to psbungee if you need.

Comment (colinrgodsey) [reply src/protocolsupport/protocol/pipeline/version/v_pe/ToClientPacketEncoder.java]: hmmm not sure im seeing it in PSbungee?

Comment (Artuto) [reply src/protocolsupport/protocol/pipeline/version/v_pe/ToClientPacketEncoder.java]: "you can copy it over to PSB"

Comment (colinrgodsey) [reply src/protocolsupport/protocol/pipeline/version/v_pe/ToClientPacketEncoder.java]: looks like PS uses the apache commands tuple, which apparently extends Map.Entry anyways lol. ill switch over to it tho, looks cleaner

**Comment by caoli5288** [comment]
OMG another huge PR
Why not split it?

**Comment by colinrgodsey** [comment]
cause master doesnt work. i literally cant make another branch and actually test my changes

File: src/protocolsupport/injector/pe/PEProxyNetworkManager.java [hunk 14]
```diff
@@ -14,7 +16,16 @@
 	protected Channel serverconnection;
 
 	@Override
-	protected void channelRead0(ChannelHandlerContext ctx, ByteBuf bytebuf) throws Exception {
+	public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+		if (msg instanceof RakNetServer.BackPressure && serverconnection != null) {
+			serverconnection.config().setAutoRead(RakNetServer.BackPressure.OFF.equals(msg));
```
Comment (caoli5288) [inline src/protocolsupport/injector/pe/PEProxyNetworkManager.java]: What's this?

Comment (colinrgodsey) [reply src/protocolsupport/injector/pe/PEProxyNetworkManager.java]: My new raknet branch has back-pressure signals (stops reading from the downstream TCP connection when the raknet frame queue hits a high watermark). keeps bungeecord from filling the heap unnecessarily when a client is a slow connection

Comment (colinrgodsey) [reply src/protocolsupport/injector/pe/PEProxyNetworkManager.java]: actually right i forgot- there is a limited form of backpressure in my master branch of RakNetServer too

**Comment by caoli5288** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
Have you tested this to be significantly effective?


    
        a#ntes-pcmail-signature-default:hover {
            text-decoration: underline;
            color: #199cff;
            cursor: pointer;
        }

        a#ntes-pcmail-signature-default:active {
            text-decoration: underline;
            color: #246fce;
            cursor: pointer;
        }
    




    On 2/19/2019 00:18，Colin Godsey<notifications@github.com> wrote： 


@colinrgodsey commented on this pull request.



In src/protocolsupport/injector/pe/PEProxyNetworkManager.java:
> @@ -14,7 +16,16 @@
 	protected Channel serverconnection;
 
 	@Override
-	protected void channelRead0(ChannelHandlerContext ctx, ByteBuf bytebuf) throws Exception {
+	public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+		if (msg instanceof RakNetServer.BackPressure && serverconnection != null) {
+			serverconnection.config().setAutoRead(RakNetServer.BackPressure.OFF.equals(msg));

actually right i forgot- there is a limited form of backpressure in my master branch of RakNetServer too

**Comment by Shevchik** [comment]
Unfortunately no form of back pressure can be effective, since the minecraft server itself can't stop writing packets. So packets will just sit somewhere (most likely in native system buffer), or will start to drop which will end up in a client disconnect due to readtimeout.

**Comment by colinrgodsey** [comment]
I'll have to dig through the network code later, but i swear there was something that limited chunk sends when the client channel wasnt confirming write futures any more. 

Either way, the backpressure behavior here should be similar to a slow normal TCP java client, and push most of the buffering onto the downstream server and the TCP stack itself instead of it just filling up the raknet stack. it definitely works for preventing slow clients from getting kicked from "too big packet loss". 

Another worse issue that was happening was that the raknet reliability layer was just getting flooded- this led to lots of resends, unnecessary backoffs, issues calculating RTT, and issues figuring out priority. This also caused lots of fun things like players on bad connections trying to join a really dense or complex region, and would eventually just start eating up heap space in bungee until the player finally decided to log off.

My new branch tracks that much better- retry rate stays almost constant during chunk loads. The new branch probably does not *need* the backflow as much, but it still helps just to keep buffering under control.

**Comment by Shevchik** [comment]
Slow java tcp client will be kicked for read timeout after some time (or after reaching netty buffer size limit), so actually current behavior is almost the same.
The too big packet loss can indeed be a problem, however disabling autoread only makes it worse, since disconnected client won't be detected as fast as it is now, and we will have a tons of buffered data in native tcp buffer, or at netty level until read timeout occurs.

As for retries, backoffs, rtt, that's the main reason why the main raknet branch doesn't have them, because it's hard to actually find a good solution for all of that. The resend is however needed due to the way how raknet works, but it should work off last received nack/ack, not fixed time.

File: src/protocolsupport/utils/netty/Decompressor.java [hunk 15]
```diff
@@ -19,33 +24,55 @@ public static Decompressor create() {
...
 
 	public void recycle() {
 		handle.recycle(this);
 	}
 
-	public static byte[] decompressStatic(byte[] input, int uncompressedlength) throws DataFormatException {
-		Decompressor decompressor = create();
+	@Override
+	public void finalize() throws Throwable {
+		zlib.free();
```
Comment (Shevchik) [inline src/protocolsupport/utils/netty/Decompressor.java]: That doesn't work. The zlib object itself can gced before wrapper object.

Comment (colinrgodsey) [reply src/protocolsupport/utils/netty/Decompressor.java]: i think im going to revert the zlib stuff in this branch. it got accidentally copied over from my dev branch

**Comment by colinrgodsey** [comment]
Disconnects should still be reasonably fast though- if the RakNet channel times out and closes it'll take the bungee tcp channel with it. We can remove the backpressure stuff from here too, my newer raknet branch doesnt really need it as much. Newer raknet branch is also much simpler with regards to resends- no backoff anymore, rtt only used for retry period. The retry period could technically be replaced using only the ACK sequence, but it'll just increase the resend rate. 

Most of this work has been done because of players using like... mobile data, while overseas. It's an edge case, but it's one thats becoming more common than I thought it would be. So reducing resends has been a big goal, and now that there's metrics available, its actually been pretty easy to test and see the actual differences some of these changes make. I understand this isn't really the goal of the main raknet branch, so im trying to keep most of these changes isolated, but in an API compatible way so people can still use the more optimized one if they choose.

**Comment by colinrgodsey** [comment]
removed the backpressure and zlib stuff

**Comment by MrPowerGamerBR** [comment]
If something explodes #blameyesdog
